### PR TITLE
Wip cxx20 fixes core

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
LGTM.

**Review by cbodley — APPROVED** [review]

**Comment by adamemerson** [comment]
> neat, i didn't see the `osd: Replace deprecated use of format_to` commit before; is that enough to resolve https://tracker.ceph.com/issues/53251?

Should be.

**Comment by adamemerson** [comment]
Looks like the make check failures are legitimate? Dumping core in basic_string.

**Comment by cbodley** [comment]
> > neat, i didn't see the `osd: Replace deprecated use of format_to` commit before; is that enough to resolve https://tracker.ceph.com/issues/53251?
> 
> Should be.

that tracker shows the same deprecation warning, but refers to `mon/LogMonitor.cc` which isn't in this diff. let's keep that tracker open

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
being tested at https://pulpito.ceph.com/kchai-2022-03-17_05:18:57-rados-wip-cxx20-fixes-core-kefu-distro-default-smithi/

File: src/include/buffer.h [hunk 1]
```diff
@@ -440,7 +440,13 @@ struct error_code;
 	buffers_iterator(U* const p)
 	  : cur(p) {
 	}
-	template <class U>
+	// copy constructor
+	buffers_iterator(const buffers_iterator<T>& other)
+	  : cur(other.cur) {
+	}
+	// converting constructor, from iterator -> const_iterator only
+	template <class U, std::enable_if_t<
```
Comment (tchaikov) [inline src/include/buffer.h:448]: i have following failure:
```
2022-03-17T05:44:34.277 INFO:tasks.workunit.client.0.smithi067.stderr:In file included from /usr/include/rados/librados.hpp:11,
2022-03-17T05:44:34.277 INFO:tasks.workunit.client.0.smithi067.stderr:                 from /home/ubuntu/cephtest/mnt.0/client.0/tmp/rocksdb/include/rocksdb/utilities/env_librados.h:14,
2022-03-17T05:44:34.278 INFO:tasks.workunit.client.0.smithi067.stderr:                 from /home/ubuntu/cephtest/mnt.0/client.0/tmp/rocksdb/utilities/env_librados.cc:4:
2022-03-17T05:44:34.278 INFO:tasks.workunit.client.0.smithi067.stderr:/usr/include/rados/buffer.h:448:26: error: ‘std::enable_if_t’ has not been declared
2022-03-17T05:44:34.278 INFO:tasks.workunit.client.0.smithi067.stderr:  template <class U, std::enable_if_t<
2022-03-17T05:44:34.278 INFO:tasks.workunit.client.0.smithi067.stderr:                          ^~~~~~~~~~~
2022-03-17T05:44:34.279 INFO:tasks.workunit.client.0.smithi067.stderr:/usr/include/rados/buffer.h:448:37: error: expected ‘>’ before ‘<’ token
2022-03-17T05:44:34.279 INFO:tasks.workunit.client.0.smithi067.stderr:  template <class U, std::enable_if_t<
2022-03-17T05:44:34.279 INFO:tasks.workunit.client.0.smithi067.stderr:                                     ^
2022-03-17T05:44:34.282 INFO:tasks.workunit.client.0.smithi067.stderr:/usr/include/rados/buffer.h: In member function ‘void ceph::buffer::v15_2_0::list::buffers_t::insert_after(ceph::buffer::v15_2_0::list::buffers_t::const_iterator, ceph::buffer::v15_2_0::list::buffers_t::reference)’:
```

see https://pulpito.ceph.com/kchai-2022-03-17_05:18:57-rados-wip-cxx20-fixes-core-kefu-distro-default-smithi/6740698/

Comment (tchaikov) [reply src/include/buffer.h:448]: as rocksdb is still a C++11 project. we have to stick with `std::enable_if<>`.

**Comment by tchaikov** [comment]
will rerun the failed test once https://shaman.ceph.com/builds/ceph/wip-cxx20-fixes-core-kefu/03b299e4f8dc02c1755bcaa0e51c76a8c250cb32/ is completed.

**Comment by tchaikov** [comment]
jenkins test make check

File: src/include/buffer.h (hunk 1 above)
Comment (cbodley) [reply src/include/buffer.h]: is it possible to enable c++20 just for `env_librados.cc`?

Comment (cbodley) [reply src/include/buffer.h]: on second thought, i agree that it's better to support c++11 librados clients

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2022-03-17_13:05:53-rados-wip-cxx20-fixes-core-kefu-distro-default-smithi/

File: src/include/buffer.h (hunk 1 above)
Comment (tchaikov) [reply src/include/buffer.h]: indeed, rocksdb is just one of them.

Comment (tchaikov) [reply src/include/buffer.h]: @cbodley i am going to squash the fix-up change into your commit if it's fine by you.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

File: src/include/buffer.h (hunk 1 above)
Comment (tchaikov) [reply src/include/buffer.h]: i was wrong. turns out rocksdb switched to C++17 in https://github.com/facebook/rocksdb/commit/fd3e0f43b3cd041c38e37543e0b616d18ca55179.

**Pull request merged**